dango/dex/src/core/liquidity_pool.rs

```diff
@@ -114,6 +160,94 @@ impl PassiveLiquidityPool for PairParams {
 
         Ok((reserve, refund))
     }
+
+    fn swap_exact_amount_in(
+        &self,
+        reserve: CoinPair,
+        input: Coin,
+    ) -> anyhow::Result<(CoinPair, Coin)> {
+        let mut new_reserve = reserve.clone();
+
+        let output_denom = if reserve.first().denom == &input.denom {
+            reserve.second().denom.clone()
+        } else {
+            reserve.first().denom.clone()
+        };
+
+        let a = reserve.amount_of(&input.denom)?;
+        let b = reserve.amount_of(&output_denom)?;
+        let out_amount = match self.curve_invariant {
+            CurveInvariant::Xyk => {
+                // Solve A * B = (A + offer.amount) * (B - amount_out) for amount_out
+                // => amount_out = B - (A * B) / (A + offer.amount)
+                // Round so that user takes the loss
+                let amount_out = b.checked_sub(
+                    Uint128::ONE.checked_multiply_ratio_ceil(a * b, a + input.amount)?,
+                )?;
```

Comment:
1. It's better we use checked math instead `*` and `-`
2. Why not this:

```diff
  let amount_out = b.checked_sub(
-    Uint128::ONE.checked_multiply_ratio_ceil(a * b, a + input.amount)?,
+    a.checked_multiply_ratio(b, a.checked_add(input.amount)?)?,
  )?;
```

Reply:
I made this change and all tests pass. I think they are equivalent.

Reply:
yes should be equivalent

Reply:
that's definitely better. don't know why I didn't use checked math in that instance